src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+        status = STATUS_NO_MEMORY;
+        goto Cleanup;
+    }
+
+    //calculate the length of the hash
+    if (!NT_SUCCESS(status = BCryptGetProperty(
+        hAlg,
+        BCRYPT_HASH_LENGTH,
+        (PBYTE)&cbHash,
+        sizeof(DWORD),
+        &cbData,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptGetProperty\n";
+        goto Cleanup;
+    }
+
+    //allocate the hash buffer on the heap
+    pbHash = (PBYTE)HeapAlloc(GetProcessHeap(), 0, cbHash);
+    if (NULL == pbHash)
+    {
+        errorString = "Memory allocation failed\n";
+        status = STATUS_NO_MEMORY;
+        goto Cleanup;
+    }
+
+    //create a hash
+    if (!NT_SUCCESS(status = BCryptCreateHash(
+        hAlg,
+        &hHash,
+        pbHashObject,
+        cbHashObject,
+        NULL,
+        0,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptCreateHash\n";
+        goto Cleanup;
+    }
+
+
+    //hash some data
+    if (!NT_SUCCESS(status = BCryptHashData(
+        hHash,
+        (PBYTE)data,
+        (ULONG)data_size,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptHashData\n";
+        goto Cleanup;
+    }
+
+    //close the hash
+    if (!NT_SUCCESS(status = BCryptFinishHash(
+        hHash,
+        pbHash,
+        cbHash,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptFinishHash\n";
+        goto Cleanup;
+    }
+
+    retval.reserve(2 * cbHash + 1);
+    for (size_t i = 0; i < cbHash; i++)
+    {
+        char buf[3];
+        snprintf(buf, sizeof(buf), "%02x", pbHash[i]);
+        retval += buf;
+    }
+
+Cleanup:
+
+    if (hAlg)
+    {
+        BCryptCloseAlgorithmProvider(hAlg, 0);
+    }
+
+    if (hHash)
+    {
+        BCryptDestroyHash(hHash);
+    }
+
+    if (pbHashObject)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHashObject);
+    }
+
+    if (pbHash)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHash);
+    }
+
+    throw_if(!NT_SUCCESS(status), errorString);
+
+    return retval;
+}
+
+static std::string sha256(const std::string& input)
+{
+    return sha256(input.length(), input.data());
+}
+
+static std::string get_file_name(const std::string& id)
+{
+    return g_cache_dirname + "\\" + sha256(id);
+}
+
+void local_cache_clear() {
+    init();
+
+    WIN32_FIND_DATA data;
+    std::wstring baseDir(g_cache_dirname.begin(), g_cache_dirname.end());
+    std::wstring searchPattern = baseDir + L"\\*";
+
+    HANDLE hFind = FindFirstFile(searchPattern.c_str(), &data);
+
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do {
+            std::wstring fileName(data.cFileName);
+            if (fileName.compare(L".") != 0 && fileName.compare(L"..") != 0)
+            {
+                std::wstring fullFileName = baseDir + L"\\" + fileName;
+
+                throw_if(!DeleteFileW(fullFileName.c_str()),
```

Comment:
What if another instance of the cache is accessing this?

Reply:
Should have either a shared lock from read or exclusive lock from write. It will delete after those handles are closed i believe